src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -402,229 +520,317 @@ private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
             {
-                Contracts.Assert(Utils.Size(indices) >= count);
+                Host.AssertValue(input);
+                Host.Assert(0 <= iinfo && iinfo < _parent.ColumnPairs.Length);
+                int src = _cols[iinfo];
+                Host.Assert(input.IsColumnActive(src));
+                return input.GetGetter<T>(src);
+            }
 
-                int offs = 0;
-                for (int i = 0; i < cdst; i++)
+            private static void FillValues(float[] model, ref VBuffer<float> src, ref VBuffer<float> dst, int cdst)
+            {
+                int count = src.Count;
+                int length = src.Length;
+                var values = src.Values;
+                var indices = src.Indices;
+                Contracts.Assert(Utils.Size(values) >= count);
+
+                // Since the whitening process produces dense vector, always use dense representation of dst.
+                var a = Utils.Size(dst.Values) >= cdst ? dst.Values : new float[cdst];
+                if (src.IsDense)
                 {
-                    a[i] = DotProduct(model, offs, values, indices, count);
-                    offs += length;
+                    Mkl.Gemv(Mkl.Layout.RowMajor, Mkl.Transpose.NoTrans, cdst, length,
+                        1, model, length, values, 1, 0, a, 1);
                 }
+                else
+                {
+                    Contracts.Assert(Utils.Size(indices) >= count);
+
+                    int offs = 0;
+                    for (int i = 0; i < cdst; i++)
+                    {
+                        a[i] = DotProduct(model, offs, values, indices, count);
+                        offs += length;
+                    }
+                }
+                dst = new VBuffer<float>(cdst, a, dst.Indices);
+            }
+
+            /// <summary>
+            /// Returns a dot product of dense vector 'a' starting from offset 'aOffset' and sparse vector 'b'
+            /// with first 'count' valid elements and their corresponding 'indices'.
+            /// </summary>
+            private static float DotProduct(float[] a, int aOffset, float[] b, int[] indices, int count)
+            {
+                Contracts.Assert(count <= indices.Length);
+                return CpuMathUtils.DotProductSparse(a.AsSpan(aOffset), b, indices, count);
+
             }
-            dst = new VBuffer<Float>(cdst, a, dst.Indices);
         }
+    }
 
-        /// <summary>
-        /// Returns a dot product of dense vector 'a' starting from offset 'aOffset' and sparse vector 'b'
-        /// with first 'count' valid elements and their corresponding 'indices'.
-        /// </summary>
-        private static Float DotProduct(Float[] a, int aOffset, Float[] b, int[] indices, int count)
+    /// <include file='doc.xml' path='doc/members/member[@name="Whitening"]/*'/>
+    public sealed class Whitening : IEstimator<WhiteningTransform>
```

Comment:
>Whitening [](start = 24, length = 9)

I think @Zruty0 prefer to call them Estimators, WhiteningEstimator.

Reply:
Should WhiteningTransform also be renamed to something else?

---
In reply to: [229457109](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229457109) [](ancestors = 229457109)